File: auto_tune_runner.js
/* Auto-tune runner (Node)
   Runs a broader randomized search using the CCM model (from js/ccm.js) and writes the top-3 configs
   to configs/best_particle_configs.json for use in the app.

   Usage: node scripts/auto_tune_runner.js
*/
const fs = require('fs');
const path = require('path');
const vm = require('vm');

const ccmPath = path.join(__dirname, '..', 'js', 'ccm.js');
const outPath = path.join(__dirname, '..', 'configs');
if(!fs.existsSync(outPath)) fs.mkdirSync(outPath);
const outFile = path.join(outPath, 'best_particle_configs.json');

const code = fs.readFileSync(ccmPath, 'utf8');
// append a module export to get the CCM object from the evaluated code
const wrapped = `${code}\n;typeof CCM !== 'undefined' ? global.CCM = CCM : null;`;
const sandbox = {};
vm.createContext(sandbox);
vm.runInContext(wrapped, sandbox);
const CCM = sandbox.CCM;
if(!CCM) { console.error('Failed to load CCM from js/ccm.js'); process.exit(1); }

// randomized search
function runSearch(trials=5000){
  const ranges = { ap:[0.4,2.0], ps:[0.4,1.6], sa:[0.4,1.4], dur:[0.6,1.6] };
  const scenarioKeys = ['pre','present','high','recover'];
  const presets = {
    pre: {co2:280, temp:14, light:70, storms:false, pollution:false},
    present: {co2:420, temp:25, light:80, storms:false, pollution:false},
    high: {co2:950, temp:36, light:95, storms:true, pollution:true},
    recover: {co2:360, temp:22, light:85, storms:false, pollution:false}
  };

  function sample(){ const s=(a,b)=> a + Math.random()*(b-a); return {apScale: Number(s(...ranges.ap).toFixed(3)), psScale: Number(s(...ranges.ps).toFixed(3)), saScale: Number(s(...ranges.sa).toFixed(3)), durationScale: Number(s(...ranges.dur).toFixed(3))}; }

  function scoreConfig(cfg){
    const rates = [];
    for(const k of scenarioKeys){
      const sc = presets[k];
      const prate = CCM.photosynthesisRate(sc);
      const cf = CCM.carbonFlow(prate, 50);
      const pAP = Math.min(1, (cf.absorbed/22) * cfg.apScale);
      const pPS = Math.min(1, (cf.emitted/28) * cfg.psScale);
      const pSA = Math.min(1, (cf.emitted/55) * cfg.saScale);
      rates.push(pAP + pPS + pSA);
    }
    const mean = rates.reduce((a,b)=>a+b,0)/rates.length;
    const variance = rates.reduce((a,b)=>a+Math.pow(b-mean,2),0)/rates.length;
    const sd = Math.sqrt(variance);
    const meanPenalty = Math.abs(mean - 0.85);
    const saturationPenalty = Math.max(0, mean - 1.25);
    return sd - meanPenalty - saturationPenalty*1.5;
  }

  const scored = [];
  for(let i=0;i<trials;i++){
    const cfg = sample(); const s = scoreConfig(cfg);
    scored.push({cfg, score:s});
  }
  scored.sort((a,b)=>b.score - a.score);
  return scored.slice(0,3);
}

console.log('Running auto-tune search (this may take a few seconds)...');
const top3 = runSearch(8000);
fs.writeFileSync(outFile, JSON.stringify({generatedAt: new Date().toISOString(), top3}, null, 2));
console.log(`Wrote top-3 configurations to ${outFile}`);
console.log('Top 3 summary:'); top3.forEach((t,i)=> console.log(`${i+1}: score=${t.score.toFixed(3)} cfg=${JSON.stringify(t.cfg)}`));

File: auto_tune_runner.py
import json
import math
import random
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
OUT = ROOT / 'configs' / 'best_particle_configs.json'

# Reimplement the same functions as CCM

def clamp(x,a,b):
    return max(a, min(b, x))

def tempFactor(temp):
    opt = 25.0; sigma = 10.0
    f = math.exp(-((temp - opt)**2)/(2*sigma*sigma))
    return clamp(f, 0.0, 1.0)

def co2Factor(co2):
    km = 600.0
    return clamp(co2 / (co2 + km), 0.0, 1.0)

def lightFactor(lightPercent):
    return clamp(lightPercent/100.0, 0.0, 1.2)

def photosynthesisRate(params):
    co2 = params['co2']; temp = params['temp']; light = params['light']; stressEvents = params.get('stressEvents',0); pollution = params.get('pollution',False)
    base = 1.0
    t = tempFactor(temp)
    c = co2Factor(co2)
    l = lightFactor(light)
    rate = base * t * c * l
    rate *= (0.85 ** stressEvents)
    if pollution:
        rate *= 0.7
    return clamp(rate, 0.0, 2.0)

def carbonFlow(pRate, biomass):
    absorbed = pRate * biomass * 0.9
    emitted = biomass * 0.05
    return {'absorbed': absorbed, 'emitted': emitted}

# presets
PRESETS = {
    'pre': {'co2':280, 'temp':14, 'light':70, 'storms':False, 'pollution':False},
    'present': {'co2':420, 'temp':25, 'light':80, 'storms':False, 'pollution':False},
    'high': {'co2':950, 'temp':36, 'light':95, 'storms':True, 'pollution':True},
    'recover': {'co2':360, 'temp':22, 'light':85, 'storms':False, 'pollution':False}
}

# scoring similar to JS

def score_config(cfg):
    scenarioKeys = ['pre','present','high','recover']
    rates = []
    for k in scenarioKeys:
        sc = PRESETS[k]
        prate = photosynthesisRate(sc)
        cf = carbonFlow(prate, 50)
        pAP = min(1.0, (cf['absorbed']/22.0) * cfg['apScale'])
        pPS = min(1.0, (cf['emitted']/28.0) * cfg['psScale'])
        pSA = min(1.0, (cf['emitted']/55.0) * cfg['saScale'])
        rates.append(pAP + pPS + pSA)
    mean = sum(rates)/len(rates)
    sd = (sum((r-mean)**2 for r in rates)/len(rates))**0.5
    meanPenalty = abs(mean - 0.85)
    saturationPenalty = max(0.0, mean - 1.25)
    return sd - meanPenalty - saturationPenalty*1.5


def run_search(trials=15000):
    ranges = {'ap':(0.4,2.0),'ps':(0.4,1.6),'sa':(0.4,1.4),'dur':(0.6,1.6)}
    results = []
    for i in range(trials):
        cfg = {
            'apScale': round(random.uniform(*ranges['ap']), 3),
            'psScale': round(random.uniform(*ranges['ps']), 3),
            'saScale': round(random.uniform(*ranges['sa']), 3),
            'durationScale': round(random.uniform(*ranges['dur']), 3)
        }
        s = score_config(cfg)
        results.append({'cfg': cfg, 'score': s})
    results.sort(key=lambda x: x['score'], reverse=True)
    return results[:3]

if __name__ == '__main__':
    print('Running broader auto-tune search (this may take a few seconds)...')
    top3 = run_search(8000)
    payload = {'generatedAt': None, 'top3': top3}
    payload['generatedAt'] = __import__('datetime').datetime.utcnow().isoformat() + 'Z'
    with OUT.open('w', encoding='utf8') as f:
        json.dump(payload, f, indent=2)
    print(f'Wrote top-3 configs to {OUT}')
    for i,t in enumerate(top3, start=1):
        print(f"{i}: score={t['score']:.4f} cfg={t['cfg']}")
